docs/design_proposals/ko-builder.md

```diff
@@ -0,0 +1,440 @@
+
+    ```go
+    type ArtifactType struct {
+      [...]
+    	// KoArtifact builds images using [ko](https://github.com/google/ko).
+    	KoArtifact *KoArtifact `yaml:"ko,omitempty" yamltags:"oneOf=artifact"`
+    }
+    ```
+
+3.  Define `KoDependencies`:
+
+    ```go
+    // KoDependencies is used to specify dependencies for an artifact built by ko.
+    type KoDependencies struct {
+	  	// Paths should be set to the file dependencies for this artifact, so that the skaffold file watcher knows when to rebuild and perform file synchronization.
+	  	// Defaults to {"go.mod", "**.go"}
+	  	Paths []string `yaml:"paths,omitempty" yamltags:"oneOf=dependency"`
+
+	  	// Ignore specifies the paths that should be ignored by skaffold's file watcher.
+    	// If a file exists in both `paths` and in `ignore`, it will be ignored, and will be excluded from both rebuilds and file synchronization.
+	  	Ignore []string `yaml:"ignore,omitempty"`
+    }
+    ```
+
+4.  Add `KO` to the `BuilderType` enum in `proto/enums/enums.proto`:
+
+    ```proto
+    enum BuilderType {
+        // Could not determine builder type
+        UNKNOWN_BUILDER_TYPE = 0;
+        // JIB Builder
+        JIB = 1;
+        // Bazel Builder
+        BAZEL = 2;
+        // Buildpacks Builder
+        BUILDPACKS = 3;
+        // Custom Builder
+        CUSTOM = 4;
+        // Kaniko Builder
+        KANIKO = 5;
+        // Docker Builder
+        DOCKER = 6;
+        // Ko Builder
+        KO = 7;
+    }
+    ```
+
+5.  In `skaffold init`, default to the ko builder for any images where the
+    name starts with the ko prefix `ko://`.
+
+### Builder config schema
+
+Example basic config, this will be sufficient for many users:
+
+```yaml
+apiVersion: skaffold/v2beta15
+kind: Config
+build:
+  artifacts:
+  - image: ko://github.com/GoogleContainerTools/skaffold/examples/ko
+    ko: {}
+```
+
+The value of the `image` field is the Go import path of the app entry point,
+[prefixed by `ko://`](https://github.com/google/ko/pull/58).
+
+A more comprehensive example config:
+
+```yaml
+apiVersion: skaffold/v2beta15
+kind: Config
+build:
+  artifacts:
+  - image: ko://github.com/GoogleContainerTools/skaffold/examples/ko-complete
+    ko:
+      asmflags: []
+      fromImage: gcr.io/distroless/static-debian10:nonroot
+      dependencies:
+        paths:
+        - go.mod
+        - "**.go"
+      env: []
+      args:
+      - -trimpath
+      - -v
+      gcflags:
+      - -m
+      labels:
+        foo: bar
+        baz: frob
+      ldflags:
+      - -buildid=
+      - -s
+      - -w
+      platforms:
+      - linux/amd64
+      - linux/arm64
+```
+
+ko requires setting a
+[`KO_DOCKER_REPO`](https://github.com/google/ko#choose-destination)
+environment variable to specify where container images are pushed. The Skaffold
+[default repo](https://skaffold.dev/docs/environment/image-registries/)
+maps directly to this value.
+
+### Open questions
```

Comment:
How can `skaffold debug` identify Ko images as being Go?  Is there an image label or environment variable set?

Reply:
Can `skaffold debug` look at the artifact type in `skaffold.yaml`?

If not, I'd suggest looking for the [`KO_DATA_PATH`](https://github.com/google/ko#static-assets) environment variable, which is always set.

Other options:
- look at the entrypoint, it's always `/ko-app/ko`
- check if `PATH` contains `/ko-app`
- look at the `author` field of the image layer at the top of the stack (in the image configuration), the value will be `ko`.